### DISCUSSION TITLE
question on valid/allowed edges

### DISCUSSION THREAD

**Discussion by nilsnolde** [valhalla/valhalla#3247] [Q&A] [answered]
I'm wondering about the following statement:

https://github.com/valhalla/valhalla/blob/eb6c7fbe7dc0561e269b88c4de1051eb81d7d0b5/src/loki/polygon_search.cc#L156-L161

wouldn't that skip one-way streets and other cases where only one direction is allowed or has a valid graph id? it's using this `Allowed` version:
https://github.com/valhalla/valhalla/blob/eb6c7fbe7dc0561e269b88c4de1051eb81d7d0b5/valhalla/sif/dynamiccost.h#L223-L235

so, if an edge _or_ its opposing is not forward accessible, both directional edges are skipped from `polygon_search` right? I'm quite sure I'm missing smth though..

and one more question: when (which circumstances) does an edge ID get a `kInvalidGraphId`?

a few hints would save me a couple of hours of researching:)

EDIT: some background: I wanted to re-use the `polygon_search` method while working on #3236 and there it'd be important that it doesn't skip both directions if only one is not valid/allowed. hence the questions.

**Comment by nilsnolde** [comment]
Uff, forget about most of it, stupid question actually.. 

however, it'd be interesting still under which circumstance a GraphId stays a `kInvalidGraphId`.. seems pretty involved to research by myself..

**Reply by kevinkreiser** [reply]
the circumstances underwhich this happens is when you have a tile set you are workign with that isnt the **whole** tile set. so say you build the data for north america but you only have pennsylvania on your device. in that case, when you get to a tile that is on the edge of pennsylvania and you have an edge which leaves that last tile on the edge and terminates in a tile that you dont have, then  you will also not have the opposing edge, because that opposing edge begins in a tile that you dont have. so the code wil know where to go to look for it but when it checks the filesystem or whatever it wont find the tile and in that case it will return kInvalidGraphId to signal its not available.

**Reply by nilsnolde** [reply]
ah perfect explanation, thanks a bunch @kevinkreiser ! So, one-ways don't have a `kInvalidGraphId` on their opposite edge right? that's flagged by `reverseaccess()` bit field I guess?

**Reply by kevinkreiser** [reply]
that is correct, they dont. they have a valid opposing edge, which other modes of travel can use, like walking the wrong way down a oneway road for example or which cars can use when you use the costing option that disables caring about oneway-ness. and yes its the access which allows to tell the traversability of the edge from the point of view of the costing. invalid graph id is just a way to say i dont have the data you need, meaning i know where your opposing edge should be in the data but its not there (the tile is missing from your tileset).

**Comment by kevinkreiser** [comment, marked as answer]
to answer the larger question yes they all do have opposing edges. though i do remember some trickery for the edges that make up transit lines in the transit tiles, i kind of forget how that works but i vaguely remember that those didnt have opposing edges. in general though yes all edges have an opposing edge running in the opposite direction (reversed begin and end nodes)
